### ISSUE TITLE
Upgrade to 3.5: Pod is completed but workflow is running?

### ISSUE THREAD

**Issue by zhucan** [argoproj/argo-workflows#13335]
### Pre-requisites

- [X] I have double-checked my configuration
- [ ] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [ ] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

Pod is completed but workflow is running， but i expect pod is completed and workflow is succeeded

```
root@10-16-10-122:/home/devops# kubectl get wf  -n argo-data-closeloop prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r 
NAME                                                              STATUS    AGE   MESSAGE
prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r   Running   23h   
root@10-16-10-122:/home/devops# kubectl get pods -n argo-data-closeloop | grep prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-job-entrypoint-217710983    0/2     Completed                  0          23h
```

### Version(s)

v3.5.8

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
because of used the private, not easy to reproduces
```

### Logs from the workflow controller

```text
root@10-16-10-122:/home/devops# kubectl logs -n argo-workflow deploy/argo-workflow-workflow-controller | grep prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:50:49.589Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=788769749 namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:50:49.590Z" level=info msg="Task-result reconciliation" namespace=argo-data-closeloop numObjs=1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:50:49.590Z" level=info msg="task-result changed" namespace=argo-data-closeloop nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5-217710983 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:50:49.590Z" level=info msg="node changed" namespace=argo-data-closeloop new.message= new.phase=Succeeded new.progress=0/1 nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-217710983 old.message= old.phase=Pending old.progress=0/1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:50:49.590Z" level=info msg="pod reconciliation didn't complete, will retry" namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:50:49.599Z" level=info msg="Workflow update successful" namespace=argo-data-closeloop phase=Running resourceVersion=788769749 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:51:49.591Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=788769749 namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:51:49.592Z" level=info msg="Task-result reconciliation" namespace=argo-data-closeloop numObjs=1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:51:49.592Z" level=info msg="task-result changed" namespace=argo-data-closeloop nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5-217710983 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:51:49.592Z" level=info msg="node changed" namespace=argo-data-closeloop new.message= new.phase=Succeeded new.progress=0/1 nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-217710983 old.message= old.phase=Pending old.progress=0/1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:51:49.592Z" level=info msg="pod reconciliation didn't complete, will retry" namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:51:50.862Z" level=info msg="Workflow update successful" namespace=argo-data-closeloop phase=Running resourceVersion=788769749 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:52:49.594Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=788769749 namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:52:49.594Z" level=info msg="Task-result reconciliation" namespace=argo-data-closeloop numObjs=1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:52:49.594Z" level=info msg="task-result changed" namespace=argo-data-closeloop nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5-217710983 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:52:49.594Z" level=info msg="node changed" namespace=argo-data-closeloop new.message= new.phase=Succeeded new.progress=0/1 nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-217710983 old.message= old.phase=Pending old.progress=0/1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:52:49.595Z" level=info msg="pod reconciliation didn't complete, will retry" namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:52:49.607Z" level=info msg="Workflow update successful" namespace=argo-data-closeloop phase=Running resourceVersion=788769749 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:53:49.595Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=788769749 namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:53:49.596Z" level=info msg="Task-result reconciliation" namespace=argo-data-closeloop numObjs=1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:53:49.596Z" level=info msg="task-result changed" namespace=argo-data-closeloop nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5-217710983 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:53:49.597Z" level=info msg="node changed" namespace=argo-data-closeloop new.message= new.phase=Succeeded new.progress=0/1 nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-217710983 old.message= old.phase=Pending old.progress=0/1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:53:49.597Z" level=info msg="pod reconciliation didn't complete, will retry" namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:53:49.604Z" level=info msg="Workflow update successful" namespace=argo-data-closeloop phase=Running resourceVersion=788769749 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:54:49.598Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=788769749 namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:54:49.598Z" level=info msg="Task-result reconciliation" namespace=argo-data-closeloop numObjs=1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:54:49.598Z" level=info msg="task-result changed" namespace=argo-data-closeloop nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5-217710983 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:54:49.598Z" level=info msg="node changed" namespace=argo-data-closeloop new.message= new.phase=Succeeded new.progress=0/1 nodeID=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-217710983 old.message= old.phase=Pending old.progress=0/1 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:54:49.598Z" level=info msg="pod reconciliation didn't complete, will retry" namespace=argo-data-closeloop workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
time="2024-07-11T07:54:49.608Z" level=info msg="Workflow update successful" namespace=argo-data-closeloop phase=Running resourceVersion=788769749 workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r
```

### Logs from in your workflow's wait container

```text
root@10-16-10-122:/home/devops# kubectl logs -n argo-data-closeloop  -c wait -l workflows.argoproj.io/workflow=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r,workflow.argoproj.io/phase!=Succeeded
time="2024-07-10T08:06:05.723Z" level=info msg="Getting file from s3" bucket=argo-workflow-artifacts endpoint="s3.deeproute.cn:80" key=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r/prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-job-entrypoint-217710983/main.log path=/tmp/argo/outputs/logs/main.log.tmp
time="2024-07-10T08:06:05.768Z" level=info msg="Load artifact" artifactName=main-logs duration=45.922895ms error="The specified key does not exist." key=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r/prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-job-entrypoint-217710983/main.log
time="2024-07-10T08:06:05.769Z" level=info msg="S3 Save path: /tmp/argo/outputs/logs/main.log, key: prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r/prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-job-entrypoint-217710983/main.log"
time="2024-07-10T08:06:05.769Z" level=info msg="Creating minio client using static credentials" endpoint="s3.deeproute.cn:80"
time="2024-07-10T08:06:05.769Z" level=info msg="Saving file to s3" bucket=argo-workflow-artifacts endpoint="s3.deeproute.cn:80" key=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r/prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-job-entrypoint-217710983/main.log path=/tmp/argo/outputs/logs/main.log
time="2024-07-10T08:06:05.879Z" level=info msg="Save artifact" artifactName=main-logs duration=110.551656ms error="<nil>" key=prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r/prod--prod--lt-submit-manual-annotation-task-1-1-51--f0529fsr5r-job-entrypoint-217710983/main.log
time="2024-07-10T08:06:05.879Z" level=info msg="not deleting local artifact" localArtPath=/tmp/argo/outputs/logs/main.log
time="2024-07-10T08:06:05.879Z" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
time="2024-07-10T08:06:05.892Z" level=info msg="Alloc=11918 TotalAlloc=19410 Sys=34173 NumGC=4 Goroutines=12"
time="2024-07-10T08:06:05.892Z" level=info msg="Deadline monitor stopped"
```

**Comment by zhucan** [comment]
I know how to reproduce it; because of the wf was created by argo controller(older version is v3.4.9),  before the wf was deleted, upgrade the version to v3.5.8.
https://github.com/argoproj/argo-workflows/blob/c59ff53a37a7e7ac41f73b23e621f601eab9d604/workflow/controller/taskresult.go#L67

updates:
```
		// If the task result is completed, set the state to true.
		if _, ok := result.Labels[common.LabelKeyReportOutputsCompleted]; !ok || result.Labels[common.LabelKeyReportOutputsCompleted] == "true" {
			woc.log.Debugf("Marking task result complete %s", resultName)
			woc.wf.Status.MarkTaskResultComplete(resultName)
		} else {
			woc.log.Debugf("Marking task result incomplete %s", resultName)
			woc.wf.Status.MarkTaskResultIncomplete(resultName)
		}
```

**Comment by Joibel** [comment]
Should be fixed by #13332
